testing/cuda/reduce_by_key.cu

```diff
@@ -286,3 +289,103 @@ void TestReduceByKeyCudaStreamsNoSync()
 }
 DECLARE_UNITTEST(TestReduceByKeyCudaStreamsNoSync);
 
+
+// Maps indices to key ids
+class div_op : public thrust::unary_function<long long, long long>
+{
+  long long m_divisor;
+
+public:
+  __host__ div_op(long long divisor)
+    : m_divisor(divisor)
+  {}
+
+  __host__ __device__
+  long long operator()(long long x) const
+  {
+    return x / m_divisor;
+  }
+};
+
+// Produces unique sequence for key
+class mod_op : public thrust::unary_function<long long, long long>
+{
+  long long m_divisor;
+
+public:
+  __host__ mod_op(long long divisor)
+    : m_divisor(divisor)
+  {}
+
+  __host__ __device__
+  long long operator()(long long x) const
+  {
+    // div: 2          
+    // idx: 0 1   2 3   4 5 
+    // key: 0 0 | 1 1 | 2 2 
+    // mod: 0 1 | 0 1 | 0 1
+    // ret: 0 1   1 2   2 3
+    return (x % m_divisor) + (x / m_divisor);
+  }
+};
+
+
+void TestReduceByKeyWithBigIndexesHelper(int magnitude)
+{
+  const long long key_size_magnitude = 8;
+  ASSERT_EQUAL(true, key_size_magnitude < magnitude);
+
+  const long long num_items       = 1ll << magnitude;
+  const long long num_unique_keys = 1ll << key_size_magnitude;
+
+  // Size of each key group
+  const long long key_size = num_items / num_unique_keys;
+
+  using counting_it      = thrust::counting_iterator<long long>;
+  using transform_key_it = thrust::transform_iterator<div_op, counting_it>;
+  using transform_val_it = thrust::transform_iterator<mod_op, counting_it>;
+  using discard_it       = thrust::discard_iterator<>;
+
+  counting_it count_begin(0ll);
+  counting_it count_end = count_begin + num_items;
+  ASSERT_EQUAL(static_cast<long long>(thrust::distance(count_begin, count_end)),
+               num_items);
+
+  transform_key_it keys_begin(count_begin, div_op{key_size});
+  transform_key_it keys_end(count_end, div_op{key_size});
+
+  transform_val_it values_begin(count_begin, mod_op{key_size});
+  discard_it output_keys;
```

Comment:
Instead of discarding, we could write an iterator that validates the written keys. For example, see: https://github.com/NVIDIA/thrust/blob/main/testing/async/exclusive_scan/large_indices.cu#L24-L102

Reply:
Thank you for the suggestion! Custom iterator approach leads to quite a lot of code, though. Since there are only 256 unique keys, I'll just store the output keys and compare them with a counting iterator.